core/client/src/blockchain.rs

```diff
@@ -89,6 +91,123 @@ pub trait Backend<Block: BlockT>: HeaderBackend<Block> {
 
 	/// Return hashes of all blocks that are children of the block with `parent_hash`.
 	fn children(&self, parent_hash: Block::Hash) -> Result<Vec<Block::Hash>>;
+
+	/// Get the most recent block hash of the best (longest) chains
+	/// that contain block with the given `target_hash`.
+	///
+	/// The search space is always limited to blocks which are in the finalized
+	/// chain or descendents of it.
+	///
+	/// If `maybe_max_block_number` is `Some(max_block_number)`
+	/// the search is limited to block `numbers <= max_block_number`.
+	/// in other words as if there were no blocks greater `max_block_number`.
+	/// Returns `Ok(None)` if `target_hash` is not found in search space.
+	/// TODO: document time complexity of this, see [#1444](https://github.com/paritytech/substrate/issues/1444)
+	fn best_containing(
+		&self,
+		target_hash: Block::Hash,
+		maybe_max_number: Option<NumberFor<Block>>,
+		import_lock: &Mutex<()>
```

Comment:
Yep, I didn't change anything with the actual code. Where should the mutex be ideally?

Reply:
Ideally there wasn't any but instead a queue of imports to process and either would this code wait until the queue is empty _or_ it would inject a task in the queue to execute the code between imports (though this is might be problematic once we attempt to parallelize block import :thinking: ), but that requires this code part to become async...

I think we need to think about that design a bit more before touching it again. Let's stick with the mutex for now.